### DISCUSSION TITLE
Proposal for Enhancing the Delegate Syntax in C#

### DISCUSSION THREAD

**Discussion by samuelyounge** [dotnet/csharplang#8343] [Language Ideas]
Proposal Summary:
Transform the delegate keyword to prefix variable declarations, indicating that the variable is a reference to a function or method. This would allow for a more natural and comprehensible syntax while preserving all existing benefits of delegates and Func<T> constructs.

Example Syntax:

C#

```csharp
private int delegate<int, string> MyMethodDelegate = (input1, input2) => 
{
    int myReturnValue;
    do 
    {
        var x = DoSomeWork(input1);
        myReturnValue = x + DoSomeMoreWork(input2);
    } 
    while (myReturnValue < 100);
    
    return myReturnValue;
};
```

Benefits:

    Improved Readability: Aligns with standard C# variable declaration syntax, making it more intuitive.
    Enhanced Flexibility: Allows both lambda expressions and full method declarations.
    Consistent Syntax: Reduces the need for uncommon syntax styles currently used by delegate types.

This enhancement would streamline the use of high-order functions in C#, making the language more powerful and user-friendly.

Thank you for considering this proposal. I believe it could significantly improve the developer experience with delegates in C#.

May you walk with God always,
Sam

**Comment by colejohnson66** [comment]
I personally would also like to see unification between function pointers (`delegate*`) and `MulticastDelegate` derived types. If I could wave my magic wand and make "C#++", that’s one thing I’d do. Because right now, despite having the same signature, `Func<T, bool>` and `Predicate<T>` are not the same type.

Sadly, the work involved to do this with C# *now* is probably not worth it.

**Comment by timcassell** [comment]
I think rather than new syntax on top of existing delegates, this would play nicely with new struct delegates. #7785 (this is also what I was imagining in my alternative section). I'm not sure if this is do-able in only C# compiler, and would probably need to rely on the runtime implementing the delegate types, so that delegate from assembly A matches delegate with the same signature from assembly B.

**Reply by jaredpar** [reply]
Not sure you can really have `struct delegate` in .NET because of tearing concerns. Part of the construction process of delegates is ensuring that the `object` + method pointer pairing is valid. That works great in a `class` because after construction the data is immutable. For a `struct`, even a `readonly struct`, it's not because of tearing concerns. That means you'd have to move all of the validation that `object` + method pointer pairing is valid to the invocation point. Guessing that would have some negative perf implications. 

The path for value based delegates is likely through the `ref struct` route where tearing is not a concern.

**Reply by timcassell** [reply]
Yes, I'm aware of the tearing concerns and even mentioned it in the proposal. The issue could be alleviated with an `Atomic<T>` type if/when the runtime adds it. And if structural delegate becomes a runtime type, it could even enforce atomic reads/writes even without the `Atomic<T>` type if they deem it necessary.

The invoke speed would not be any slower than existing class-based delegates. The existing delegates require a null-check for the instance method invoke, and the struct delegate would also include a null/zero-check to invoke. It should actually be even faster because it doesn't require the indirection of the class instance and multi-cast.

**Reply by timcassell** [reply]
> The path for value based delegates is likely through the `ref struct` route where tearing is not a concern.

That seems not very useful, and would be better served through functional interfaces, imo. #3452

**Reply by jaredpar** [reply]
> And if structural delegate becomes a runtime type, it could even enforce atomic reads/writes even without the Atomic<T> type if they deem it necessary.

Part of the problem though is that it's not _could_ enforce but _must_ enforce. A `struct delegate` becomes a type that every language has to agree how to use. It is also viral because any type that has a member that is a `struct delegate`, must also be used with `Atomic<T>` (at least when assigning that is the `struct delegate`. That even starts asking questions like "can this be used in generics?"

Take a step back and it has a _lot_ of similarities to `ref struct`. Yes it doesn't have the fundamental limitation that it can't be in the heap, but the nature of the atomic assignments (and atomic reads) put many of the same constraints in place. 

> > The path for value based delegates is likely through the ref struct route where tearing is not a concern.

> That seems not very useful, and would be better served through functional interfaces, imo. https://github.com/dotnet/csharplang/issues/3452

Those can't capture `ref struct` members.

**Reply by timcassell** [reply]
> Part of the problem though is that it's not _could_ enforce but _must_ enforce. A `struct delegate` becomes a type that every language has to agree how to use. It is also viral because any type that has a member that is a `struct delegate`, must also be used with `Atomic<T>` (at least when assigning that is the `struct delegate`.

If the runtime enforces atomic r/w of struct delegates, it stands to reason it can enforce atomic r/w of those fields of a struct that is being copied. A struct containing a struct delegate being copied will read each struct delegate field atomically.

> That even starts asking questions like "can this be used in generics?"

I don't see why not? The runtime already special-cases struct generics, this would just be an extra-special-case.

> Take a step back and it has a _lot_ of similarities to `ref struct`. Yes it doesn't have the fundamental limitation that it can't be in the heap, but the nature of the atomic assignments (and atomic reads) put many of the same constraints in place.

Well, it's certainly not impossible, but I do see the large amount of work it would take to make it work. Which probably means it will never happen...

> Those can't capture `ref struct` members.

Yes they can, with the new `allows ref struct` anti-constraint.

**Reply by jaredpar** [reply]
> I don't see why not? The runtime already special-cases struct generics, this would just be an extra-special-case.

The runtime _sometimes_ special cases `struct` generics. There are versions of the runtime that do not. 

> Yes they can, with the new allows ref struct anti-constraint.

That only works if 

1. The closure type is a `ref struct`
2. State is explicitly passed and is a `ref struct` 

Either case, at that point they're fundamentally not different from a `ref struct` delegate.

**Reply by timcassell** [reply]
> > I don't see why not? The runtime already special-cases struct generics, this would just be an extra-special-case.
> 
> The runtime _sometimes_ special cases `struct` generics. There are versions of the runtime that do not.

Like NativeAOT? I'm sure there's a way to make it work, but I'm not going to waste brain-power on that at this point.

> > Yes they can, with the new allows ref struct anti-constraint.
> 
> That only works if
> 
> 1. The closure type is a `ref struct`
> 2. State is explicitly passed and is a `ref struct`

Of course the display type would need to be a ref struct to capture ref struct values. Everything would be in the display struct, you would not need to explicitly pass state. I believe this was all explained in the functional interfaces proposal.

> Either case, at that point they're fundamentally not different from a `ref struct` delegate.

They may be _functionally_ equivalent, but they are fundamentally different. Functional interfaces can capture state without allocating, while delegates must allocate a display class, whether it's a `ref struct` or otherwise. And delegates cannot capture `ref struct` members at all, even if the delegate itself is `ref struct`.

**Reply by jaredpar** [reply]
> Functional interfaces can capture state without allocating,

To a degree. Essentially only when used as a constrained type parameter. When used as the stated interface type then they will always allocate.

**Reply by jaredpar** [reply]
Functional interface are inline with one of the ways that's we've been thinking about value delegates. Probably more accurate to say allocation free delegates than value delegates.

C# 13 adds building blocks around `ref struct` we knew were necessary to solving this problem. Now that they're in place we can more seriously look at the solutions.

**Comment by KennethHoff** [comment]
If we're going to change the delegate syntax, then I think following the way it's done in Typescript is a much better proposal. I know I've seen such a proposal before, but I couldn't find it looking it up now, but the gist of it is this: 

```
// Old syntax
Func<int, int, bool>

// New syntax
(int, int) => bool
```

**Reply by timcassell** [reply]
I like it. That looks like it would also nicely work with named arguments (erased similar to ValueTuple element names).

```cs
(int index, int count) => bool
```

**Comment by jaredpar** [comment]
```csharp
private int delegate<int, string> MyMethodDelegate ...
```

Your proposal does not specify how this signature would be emitted into IL. Under the hood delegates are nominal types, not structural. That means a concrete type must be emitted for `MyMethodDelegate` in IL. What type would be used here? If the answer is variations of `Func` and `Action` then what happens when they can't be used? For example, `ref int delegate<int, string>` can't be represented through standard `Func` signatures.

**Reply by alrz** [reply]
Wouldn't this be the exact same type that lambda natural types use? 
```cs
//  delegate<int, void> => Action<int>
var lambda = (int a) => {}; 
```
which is currently anonymous, this would just give it a type syntax.

**Reply by jaredpar** [reply]
> which is currently anonymous, this would just give it a type syntax.

It's anonymous for a reason though. The types that are generated from such expression are not stable from release to release. Consider as a concrete example the following code: 

```csharp
void M(Span<int> span) { 
    var lambda = () => span;
} 
```

The type of `lambda` is different between .NET 8 and .NET 9. This is mostly uninteresting to customers as it's the type of a local. The 99% use case is let me evaluate it later in the method. It's also unobservable to callers from other assemblies because it's a local. So the type changing from release to release is generally fine and doesn't impact binary compat.

This proposal has syntax that applies to accessible members. That means the type is observable outside the current assembly. As such it changing from release to release would be a binary breaking change and generally not acceptable. The proposal would need to generate predictable types that don't change release to release.

**Reply by hez2010** [reply]
I believe we may need to introduce structural delegate types from the runtime/metadata directly so that a delegate type can no longer be nominal.

**Reply by sharpchen** [reply]
I don't know low-level stuff; can we share the concept of interface with delegates so they can remain nominal? I'm just curious.

**Reply by jaredpar** [reply]
> I don't know low-level stuff; can we share the concept of interface with delegates so they can remain nominal? I'm just curious.

The problem is that interfaces and delegates are nominal types. If they were structural types this proposal would be much easier. At that point the only item that matters is the signatures and then it's very easy to make type identity the same across releases. It also means it's very easy for different assemblies to agree on types. Again the only item that would matter is the signature so as long as two assemblies agreed on the signature then the instances would be convertible. 

Unfortunately delegates are nominal. That means they are a named type that is defined in a specific assembly. That means for two assemblies to have compatible delegate types they have to agree on the assembly where the delegate type is defined. That is not always possible because corelib doesn't contain, and can't possibly contain, every possible delegate type out there. It can get pretty far with `Action / Func` but that will never handle arbitrary number of type parameters, `in / out / ref`, modopts or modreqs. 

> I believe we may need to introduce structural delegate types from the runtime/metadata directly so that a delegate type can no longer be nominal.

For a feature like this to work I agree you pretty much need the runtime to introduce structural delegate types. Essentially the same way that function pointers work. Think that is unlikely to happen.

**Reply by alrz** [reply]
When extensions come around I think the compiler could implement adapters for those passed across assemblies, I believe that's something extensions will enable eventually?

**Reply by timcassell** [reply]
> When extensions come around I think the compiler could implement adapters for those passed across assemblies, I believe that's something extensions will enable eventually?

Extensions won't magically enable identity conversion between delegate types. The runtime checks the delegate's type when delegates are combined. The minimum that would be required is to allocate a delegate of the other type.

**Reply by alrz** [reply]
I didn't claim this works magically, I said "the compiler could implement adapters for those passed across assemblies", basically building delegatetypes on top of extensions where delegate<...> syntax is emitted as an interface that could be implemented externally. This doesn't need dedicated runtime support for structural delegates, rather reusing whatever codegen that extensions will compile to.

**Reply by jaredpar** [reply]
> The minimum that would be required is to allocate a delegate of the other type.

Correct and that creates a number of problems. The first is a silent allocation which is generally not great. The other problem is that it breaks identity with the original delegate. That means you could easily get into situations where you subtly break behavior that relies on identity: `lock`, event handler add / remove, etc ...

**Reply by timcassell** [reply]
That would just hide performance reductions behind nicer syntax, which I am not a fan of.

**Comment by kyoyama-kazusa** [comment]
#8004
